### PR TITLE
(CDAP-351) Unify Dataset[Input|Output]Format between Spark and MapReduce

### PR DESCRIPTION
**Pull request by chtyim**
- Introduce AbstractBatchReadableInputFormat and AbstractBatchWritableOutputFormat
  - Both MapReduce and Spark corresponding Input/Output formats extend from them
- Introduce Dataset[Input|Output]FormatProvider to unify handling of Table and FileSet usage in Batch programs

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed http://builds.cask.co/browse/CDAP-DUT3226-7

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceBatchWritableOutputFormat.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import co.cask.cdap.api.data.batch.BatchWritable;
+import co.cask.cdap.internal.app.runtime.batch.dataset.AbstractBatchWritableOutputFormat;
+import co.cask.cdap.internal.app.runtime.batch.dataset.CloseableBatchWritable;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.util.Map;
+
+/**
+ /**
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceBatchWritableOutputFormat.java]: remove line 28

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java [hunk 2]
```diff
@@ -246,18 +230,16 @@ public ExecutionSparkContext(ApplicationSpecification appSpec,
...
-        if (dataset instanceof DatasetOutputCommitter) {
-          ((DatasetOutputCommitter) dataset).onSuccess();
-        }
-        return;
+      } catch (ClassNotFoundException e) {
+        throw new DatasetInstantiationException(String.format(
+          "Cannot load input format class %s provided by dataset '%s'", outputFormatName, datasetName), e);
+      } catch (ClassCastException e) {
+        throw new DatasetInstantiationException(String.format(
+          "Input format class %s provided by dataset '%s' is not an input format", outputFormatName, datasetName), e);
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java]: exception message seems wrong.
It's not an input format class, but rather a output format name, right?
Same with the message 3 lines above it.

**Comment by anwar6953** [comment]
Besides a comment about exception message, remainder LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java]: Fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceBatchWritableOutputFormat.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceBatchWritableOutputFormat.java]: Fixed.

**Comment by chtyim** [comment]
Comments address. Since those changes only involve comment and error message, not waiting to rerun the build again. Merging.

**Pull request merged**